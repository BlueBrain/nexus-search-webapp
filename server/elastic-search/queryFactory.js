import * as Errors from "./errors";
import { to } from "../libs/async";

const DEFAULT_QUERY_BUILDER = () => Promise.resolve({});
const DEFAULT_NORMALIZER = docs => docs;
const DEFAULT_PARAMS = {
  size: null,
  from: null,
  filter: null,
  type: null,
  q: null
};

/**
 * Factory that makes fetchQuery
 *
 * @export
 * @param {object} client elastic search client instance
 * @returns {function} fetchQuery
 */
export default function queryFactory(
  client,
  index,
  queryBuilder = DEFAULT_QUERY_BUILDER,
  normalizer = DEFAULT_NORMALIZER
) {
  /**
   * Fetch the docs!
   *
   * @export
   * @param {object} query elastic search client instance
   * @returns {Promise} fetchQuery
   */
  return async function fetchQuery(query = DEFAULT_PARAMS, requestParams=DEFAULT_PARAMS, headers) {
    let error, body, docs;
    [error, body] = await to(queryBuilder(query, client, index, headers));
    if (error) { throw new Errors.QueryBuilderError(error); }
    const params = {
      size: query.size,
      from: query.from,
      index,
      type: "doc",
      body
    };
    [error, docs] = await to(client.search(params, headers));
    if (error) { throw new Errors.ElasticSearchError(error); }
    return normalizer(docs);
  };
}
