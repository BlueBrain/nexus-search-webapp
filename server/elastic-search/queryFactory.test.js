import queryFactory from "./queryFactory";

describe("queryFactory()", () => {
  it("returns a function", () => {
    expect(typeof queryFactory()).toBe("function");
  });

  describe("fetchQuery()", () => {
    it("returns a promise", async () => {
      const fakeTypesResults = {
        docs: {
          aggrigations: {
            "@types": {
              buckets: []
            }
          }
        }
      }
      expect.assertions(1);
      const fakeClient = {
        search: () => Promise.resolve(fakeTypesResults)
      };
      const fetchTypes = queryFactory(fakeClient, "my_index");
      const data = await fetchTypes();
      return expect(data).toBe(fakeTypesResults);
    });

    it("calls the es client search function", async () => {
      expect.assertions(1);
      const fakeClient = {
        search: jest.fn()
      };
      const fetchTypes = queryFactory(fakeClient, "my_index");
      await fetchTypes();
      return expect(fakeClient.search.mock.calls.length).toBe(1);
    });

    it("it calls es client search function with params index, type, and doc", async () => {
      const fakeClient = {
        search: jest.fn()
      };
      const fetchTypes = queryFactory(fakeClient, "my_index");
      await fetchTypes();
      expect(fakeClient.search.mock.calls[0][0]).toMatchSnapshot()
    });

    it("it calls the queryBuilder", async () => {
      const fakeClient = {
        search: jest.fn()
      };
      const fakeQueryBuilder = jest.fn();
      const fetchTypes = queryFactory(fakeClient, "my_index", fakeQueryBuilder);
      await fetchTypes();
      expect(fakeQueryBuilder).toBeCalled();
    });

    it("it calls the normalizer", async () => {
      const fakeClient = {
        search: jest.fn()
      };
      const fakeNormalizer = jest.fn();
      const fakeQueryBuilder = jest.fn();
      const fetchTypes = queryFactory(fakeClient, "my_index", fakeQueryBuilder, fakeNormalizer);
      await fetchTypes();
      expect(fakeNormalizer).toBeCalled();
    });

    it("throws an ElasticSearchError if search fails", async () => {
      const fakeClient = {
        search: () => Promise.reject("I exploded")
      };
      const fetchTypes = queryFactory(fakeClient, "my_index");
      expect.assertions(2);
      await expect(fetchTypes()).rejects.toBeInstanceOf(Error);
      await expect(fetchTypes()).rejects.toHaveProperty("name", "ElasticSearchError");
    })

    it("throws an QueryBuilderError if search fails", async () => {
      const fakeClient = {
        search: jest.fn()
      };
      const fakeQueryBuilder = () => Promise.reject(new Error("explode"));
      const fetchTypes = queryFactory(fakeClient, "my_index", fakeQueryBuilder);
      expect.assertions(2);
      await expect(fetchTypes()).rejects.toBeInstanceOf(Error);
      await expect(fetchTypes()).rejects.toHaveProperty("name", "QueryBuilderError");
    })
  });
});
