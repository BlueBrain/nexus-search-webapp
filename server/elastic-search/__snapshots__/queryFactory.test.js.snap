// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queryFactory() fetchQuery() it calls es client search function with params index, type, and doc 1`] = `
Object {
  "body": Object {},
  "from": null,
  "index": "my_index",
  "size": null,
  "type": "doc",
}
`;
